### ISSUE TITLE
Batch mode slow on 2015.5.2

### ISSUE THREAD

**Issue by msciciel** [saltstack/salt#25155]
Any chances for quicker batch job start (instant start in previous implementation was better than 10-30s delay) ? In addition i see that current implementation in version 2015.5.2 starts first batch in parallel, then next jobs in single mode and after few hosts crash occurs. Now batch mode is inusable but crucial for us.

<pre>
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in 
    salt_main()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 349, in salt_main
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/salt.py", line 98, in run
    for res in batch.run():
  File "/usr/lib/python2.6/site-packages/salt/cli/batch.py", line 179, in run
    active.remove(minion)
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in 
    salt_main()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 349, in salt_main
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/salt.py", line 98, in run
    for res in batch.run():
  File "/usr/lib/python2.6/site-packages/salt/cli/batch.py", line 179, in run
    active.remove(minion)
ValueError: list.remove(x): x not in list

           Salt: 2015.5.2
         Python: 2.6.6 (r266:84292, Jan 22 2014, 09:42:36)
         Jinja2: 2.7.2
       M2Crypto: 0.20.2
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 4.0.4
           Mako: Not Installed
</pre>

**Comment by msciciel** [comment]
Full example:

<pre>
salt -b 10 'node*.domain' test.ping

Executing run on ['node01.domain', 'node08.domain', 'node14.domain', 'node22.domain', 'node30.domain', 'node33.domain', 'node23.domain', 'node27.domain', 'node01.itg.domain', 'node35.domain']

retcode:
    0
node30.domain:
    True
retcode:
    0
node23.domain:
    True

Executing run on ['node34.domain', 'node07.domain']

retcode:
    0
node14.domain:
    True

Executing run on ['node20.domain']

node01.itg.domain:
    ----------
node01.domain:
    ----------
node33.domain:
    ----------
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 349, in salt_main
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/salt.py", line 98, in run
    for res in batch.run():
  File "/usr/lib/python2.6/site-packages/salt/cli/batch.py", line 179, in run
    active.remove(minion)
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.6/site-packages/salt/scripts.py", line 349, in salt_main
    client.run()
  File "/usr/lib/python2.6/site-packages/salt/cli/salt.py", line 98, in run
    for res in batch.run():
  File "/usr/lib/python2.6/site-packages/salt/cli/batch.py", line 179, in run
    active.remove(minion)
ValueError: list.remove(x): x not in list
</pre>

**Comment by jfindlay** [comment]
@msciciel, thanks for the report.  This is a duplicate of #24875, which has been fixed in 2015.5.3, which will come out in the next few days.  I don't have any experience with batch being slow to start.  If you want we can change the title and update the content of this issue to focus on that and I'll remove the duplicate label, etc.

**Comment by msciciel** [comment]
Correct me if i am wrong. For short time batch implementation was changed to use list of minions keys as a source of all minion instead of ping of all minions. This solution was working great for me and batch job started instantly. Now i think implementation has been reverted again to ping of all minion to fetch list of minion and in my opinion it's slow now. It requires a lot of seconds to start instead of instant start in previous implementation.

**Comment by jfindlay** [comment]
@msciciel, it sounds like you know more about this than I do.  A new command line flag could be added for batch commands to toggle this behavior.

**Comment by msciciel** [comment]
@jfindlay it would be great :)

**Comment by msciciel** [comment]
After upgrade to 2015.5.3 batch mode works better but weird( or maybe I dont understand something) . Chunk size is from 1 to batch size and it's random for example exucution is done in order (salt -b 10 '*' test.ping):
- 3 hosts
- 7 hosts
- 2 hosts
- 10 hosts
- 3 hosts

Batch mode still needs 10-30 seconds to start and return a lot o minions with no response. It's still far away from working as good as with CKminions implementation.

**Comment by msciciel** [comment]
Any updates on this topic ?

**Comment by jfindlay** [comment]
@msciciel, I remember some batch work being done recently, but I think most or all of it went into 2015.8, so you could try that.  Otherwise, I am unaware of any updates to batch.  SaltStack resources are limited at this point, meaning that most bugs are not getting attention, but we're working on new ways to create more community involvement.
